tests/ssl/test_cert_wrapper.py

```diff
@@ -0,0 +1,37 @@
+'''
+This Source Code Form is subject to the terms of the
+Mozilla Public License, v. 2.0. If a copy of the MPL
+was not distributed with this file, You can obtain one 
+at http://mozilla.org/MPL/2.0/.
+'''
+
+import datetime
+import mock
+import unittest
+
+from ..helpers import load_fixture_bytes
+
+from ignition.ssl.cert_wrapper import CertWrapper
+
+
+class CertWrapperTests(unittest.TestCase):
+  def test_certificate_parse(self):
+    cert_wrapper = CertWrapper.parse(load_fixture_bytes('sample_cert.der'))
+
+    self.assertEqual(
+      cert_wrapper.certificate.subject.rfc4514_string(), 
+      '1.2.840.113549.1.9.1=n/a,CN=n/a,OU=Test Certficate,O=Test Certificate,L=New York,ST=NY,C=US'
+    )
+    self.assertEqual(cert_wrapper.expiration(), datetime.datetime(2022, 2, 20, 17, 50, 54))
+    self.assertEqual(
+      cert_wrapper.fingerprint(), 
+      'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDzR2vptlPcQUlDHLanY9RsSwLgFHT3ecCyUicbY4HBOvlqRRL9l' +
+      'dPQGFcblws0XsHcTNJABCe4xjPhzG4kFY/27hcgvnWiVX7gW1ylqWf6+dP7m+ewkU3xKK/XVmy0hBAIxuQoUhFzSA' +
+      'zp+mx7lz3tm9OtOs2QFZu0U5FghVghSt6lUWJFDl2kWxzXxfscnCD9wFt5Plc9joxWF5xFBdx4Ocj/qpapdaiVoev' +
+      'sGYNNK5wBisG0ejKPjmGo6MLQYMslRvwRajASMNVezDPddgrYsDLcCHCK0pijA2MMKnZZlIVRgin+DSVGaf4lYwPK' +
+      'Wlf8Q6Ur6QRCn7MHaDkEFMAMvISkuQbaCJCTzZmWYKUxPolKWH+YoOC5YmhpVMJtR0naRqvJbGL21pWg8R4MXCjQM' +
+      'ObHcMZEql5gEMxyJabWuS98HuCNaf3SzVQgpI3CI3PVxmHflZRwiwDuUWTXJ0UsFPelN1Vz0XVz8148wYUxGyABnI' +
+      'RrAF0Z1htkb3iBFFo2R4MaPwLQtqNXQCdDlUXKHa4bVSQ6B7VBrL5KeDsFuvaWK+gQ9bfnqT+YSWGTVC2RyXtuvR+' +
+      'Ee1JFqQckE+x2FGbbp5ZCPl1PdgQqPg5M+vXuLdnXk3T1R6ujQTMYL42gAUktHkApLcOQQ7wYFCB1KNXYxoNGsA0p' +
+      'tMs0fsgcGQ=='
+    )
```

Comment:
Any reason you chose to use the SSH-formatted public key, over the SHA-256 fingerprint of the whole certificate? If I'm correct, this is only based on the public key in the subject, so fields in the certificate could change, and this would still be valid. Benefit of doing it this way, is you can continue using the same key, even after the certificate expires, but there's no way to expire the key itself. On mobile, so can't verify how the backend serializes the "fingerprint," but this just looks like an SSH public key.

Reply:
Yes--you're correct, this is using the public key.  This was designed for ease of serialization, but you're right that it's only based on the public key.  I think you raise a really good concern.

I'm going to do a bit more research this week on this topic and see if there is a better way to solve this so it's serializing based off the full cert, if so I'll open a separate issue for this (as this wasn't changed as a part of this PR, just moved around).

FYI--The method that does this is in ignition/ssl/cert_wrapper.py, here is the method:

```python
  def fingerprint(self) -> str:
    '''
    Extracts the public key & expiration date from the cert,
    and returns the public key openssh fingerprint
    '''
    return self.certificate.public_key().public_bytes(
      cryptography.hazmat.primitives.serialization.Encoding.OpenSSH,
      cryptography.hazmat.primitives.serialization.PublicFormat.OpenSSH
    ).decode('utf-8')
```

Reply:
If you look at [this certificate for gemini.circumlunar.space](https://crt.sh/?id=3846770529), you can see both the SHA1 and SHA256 fingerprints. These are simply the SHA1 or SHA256 hash of the DER-encoded certificate (that's the binary representation, as opposed to the PEM text). It looks like cryptography.x509.Certificate exposes a fingerprint method, that you just have to provide an algorithm to.

The disadvantage of using the fingerprint of the certificate is that in your current TOFU model, servers would have to replace their certificate *right* when the certificate expires, otherwise, people will either get a TOFU error, or an error because of an expired certificate. On the other hand, using the public key removes the ability for keys to expire at all: if I steal your key, I can simply mint a new certificate with that key that expires whenever I want it to. I personally, am not a fan of the TOFU model, but that's what we have to work it...

Reply:
By your TOFU model, I just mean that there might be another way to handle replacing existing certificates with new ones, without causing some of the issues I mentioned. I haven't researched all the ways other client authors are handling TOFU.

Reply:
The limitations you describe are real, and they are absolutely correct.  It seems like originally, some of the TOFU implementations did implement their certificate approvals this way, but have since moved away from this given the public key attack vector.  For example: https://developer.mozilla.org/en-US/docs/Web/HTTP/Public_Key_Pinning

A more detailed approach to the TOFU model is outlined here: https://rp.delaat.net/2012-2013/p56/report.pdf (captured from the mailing list).  This agrees with you that the public key pinning strategy is not really the best approach.

> In case of X.509 certificates the raw Subject Public Key Info (SPKI) structure in DER format should be pinned, which contains the type of public key and data specific to that public key type. Pinning just the public key without the type would allow attacks using a different type of public key, while pinning the whole certificate would result in more change notifications then necessary.

In order to do this, I could enhance the storage strategy to meet some of the requirements in this document, which could then be used to implement more advanced certificate approval logic.  In addition, it's probably advised to implement the verification algorithms outlined here.

I agree with your assessment that TOFU has a lot of drawbacks in practice; which is why I am interested in implementing alternative certificate verification methods if the user ultimately wants to use them.  I don't think requiring TOFU for all connections is necessarily a requirement for Gemini; even though the community seems to have gone down that path.  I think there's some flexibility here we can add.

Here's solderpunk's thoughts on this as well: https://lists.orbitalfox.eu/archives/gemini/2020/001800.html.

By the way--I'd like to move this discussion into a new issue so that we don't lose this when this PR is merged, since it's a slightly different issue than this PR addresses.  I'll open a ticket, copy this conversation over there so we don't lose it.